sig-scheduling/periodic-kueue-test-e2e-kueueviz-main

<!--
Please use this template for reporting flaky tests.
Links to specific failures in Prow are appreciated.
-->

**Which test is flaking?**:
 sig-scheduling/periodic-kueue-test-e2e-kueueviz-main

**Link to failed CI job or steps to reproduce locally**:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/periodic-kueue-test-e2e-kueueviz-main/2063748516794077184
**Failure message or logs**:
```
+ /home/prow/go/src/kubernetes-sigs/kueue/hack/testing/../../bin/kind export logs /logs/artifacts --name kind
Exporting logs for cluster "kind" to:
/logs/artifacts
+ kubectl_args=()
+ local -a kubectl_args
+ [[ -n '' ]]
+ kubectl version --request-timeout=30s
+ kubectl describe pods -n kueue-system
+ kubectl describe pods
+ cluster_cleanup kind
+ local cluster_name=kind
+ local max_retries=5
+ local retry_delay=1
++ seq 1 5
+ for attempt in $(seq 1 "$max_retries")
+ local output
++ /home/prow/go/src/kubernetes-sigs/kueue/hack/testing/../../bin/kind delete cluster --name kind
+ output='Deleting cluster "kind" ...
Deleted nodes: ["kind-worker2" "kind-worker" "kind-control-plane"]'
+ echo 'Deleting cluster "kind" ...
Deleted nodes: ["kind-worker2" "kind-worker" "kind-control-plane"]'
Deleting cluster "kind" ...
Deleted nodes: ["kind-worker2" "kind-worker" "kind-control-plane"]
+ return 0
make: *** [Makefile-test.mk:694: test-e2e-kueueviz] Error 152
+ EXIT_VALUE=2
+ set +o xtrace
Cleaning up after docker in docker.
```

**Anything else we need to know?**:

--- a/hack/testing/e2e-kueueviz-frontend.sh
+++ b/hack/testing/e2e-kueueviz-frontend.sh
@@ -22,8 +22,13 @@ if [ -n "${CI}" ] && [ -n "${PROW_JOB_ID}" ]; then
   echo "Running in prow: Disabling color to have readable logs: NO_COLOR=${NO_COLOR}"
 fi
 
+# Retry network-dependent install steps to tolerate transient registry/network
+# failures (e.g. ECONNRESET) that would otherwise flake the e2e job.
+RETRY="${PROJECT_DIR}/hack/testing/retry.sh"
+
 # Install missing dependencies
-apt-get update && apt-get install -y curl make xdg-utils
+"${RETRY}" --attempts 5 --delay 5 --stream --exponential -- apt-get update
+"${RETRY}" --attempts 5 --delay 5 --stream --exponential -- apt-get install -y curl make xdg-utils
 
 # Function to clean up background processes
 cleanup() {
@@ -33,7 +38,7 @@ cleanup() {
 
 # Run frontend unit tests before starting the app
 cd "${PROJECT_DIR}/cmd/kueueviz/frontend"
-npm install
+"${RETRY}" --attempts 5 --delay 5 --stream --exponential -- npm install
 npm test
 
 # Start kueueviz frontend
@@ -41,8 +46,8 @@ npm start & FRONTEND_PID=$!
 
 # Run Cypress tests for kueueviz frontend
 cd "${PROJECT_DIR}/test/e2e/kueueviz"
-npm install
-npx cypress install
+"${RETRY}" --attempts 5 --delay 5 --stream --exponential -- npm install
+"${RETRY}" --attempts 5 --delay 5 --stream --exponential -- npx cypress install
 npm run cypress:run --headless --config-file cypress.config.js
 
 # The trap will handle cleanup 

--- a/hack/testing/retry.sh
+++ b/hack/testing/retry.sh
@@ -17,13 +17,21 @@
 # echoes it; stderr passes through. Designed to be invoked from Make $(shell ...)
 # where the captured stdout becomes the variable value.
 #
+# By default stdout is captured and echoed only on success, so failed-attempt
+# output never pollutes a $(shell ...) caller that captures the result. Use
+# --stream when you instead want the command's stdout/stderr forwarded live
+# (e.g. for installs whose progress should be visible); --stream forwards output
+# from every attempt and returns nothing to capture, so do not combine it with a
+# $(shell ...) consumer.
+#
 # Advanced features:
 #   --exponential:       Double the delay after each failed attempt.
+#   --stream:            Forward stdout/stderr live instead of capturing stdout.
 #   --continue-if "CMD": Evaluates CMD upon failure. If it fails, retries are aborted immediately (fail-fast).
 #   --cleanup "CMD": Evaluates CMD before starting the next retry attempt.
 #
 # Usage:
-#   retry.sh [--attempts N] [--delay SECONDS] [--exponential] [--continue-if "CMD"] [--cleanup "CMD"] -- <command> [args...]
+#   retry.sh [--attempts N] [--delay SECONDS] [--exponential] [--stream] [--continue-if "CMD"] [--cleanup "CMD"] -- <command> [args...]
 #
 # Defaults: --attempts 4, --delay 5 (no exponential backoff)
 # Exit:     0 on first success, 1 if all attempts fail, 2 on usage error.
@@ -33,6 +41,7 @@ set -u
 attempts=4
 delay=5
 exponential=false
+stream=false
 continue_if=""
 cleanup=""
 
@@ -41,6 +50,7 @@ while [ $# -gt 0 ]; do
         --attempts)        attempts=$2; shift 2 ;;
         --delay)           delay=$2;    shift 2 ;;
         --exponential)     exponential=true; shift ;;
+        --stream)          stream=true; shift ;;
         --continue-if)     continue_if=$2; shift 2 ;;
         --cleanup)         cleanup=$2; shift 2 ;;
         --)                shift; break ;;
@@ -56,7 +66,14 @@ fi
 
 for i in $(seq 1 "$attempts"); do
     echo "retry [$i/$attempts]: $*" 1>&2
-    if out=$("$@"); then
+    if [ "$stream" = "true" ]; then
+        # Forward stdout/stderr live; nothing is captured or echoed.
+        if "$@"; then
+            exit 0
+        fi
+    elif out=$("$@"); then
+        # Capture stdout and echo it only on success, so failed-attempt output
+        # never pollutes a $(shell ...) caller that captures the result.
         printf '%s' "$out"
         exit 0
     fi

